cmd/limits.go

```diff
@@ -28,94 +45,140 @@ var limitsCmd = &cobra.Command{
 Supported output template annotations: %s`,
 		strings.Join(outputterTemplateAnnotations(&LimitsOutput{}), ", ")),
 	RunE: func(cmd *cobra.Command, args []string) error {
-		return output(listLimits())
-	},
-}
+		var curUsage sync.Map
+
+		// Global resources ///////////////////////////////////////////////
 
-func listLimits() (outputter, error) {
-	limits, err := cs.ListWithContext(gContext, &egoscale.ResourceLimit{})
-	if err != nil {
-		return nil, err
-	}
+		res, err := cs.RequestWithContext(gContext, &egoscale.ListAPIKeys{})
+		if err != nil {
+			return fmt.Errorf("unable to list IAM API keys: %s", err)
+		}
+		curUsage.Store(limitIAMAPIKeys, res.(*egoscale.ListAPIKeysResponse).Count)
+
+		res, err = cs.RequestWithContext(gContext, egoscale.ListBucketsUsage{})
+		if err != nil {
+			return fmt.Errorf("unable to list SOS buckets: %s", err)
+		}
+		curUsage.Store(limitSOSBuckets, res.(*egoscale.ListBucketsUsageResponse).Count)
 
-	display := map[string]string{
-		"user_vm":           "Instances",
-		"snapshot":          "Snapshots",
-		"template":          "Templates",
-		"public_elastic_ip": "IP Addresses",
-		"network":           "Private Networks",
-	}
+		// Zone-local resources /////////////////////////////////////////////
 
-	out := LimitsOutput{}
+		instanceTypes := make(map[string]*exov2.InstanceType) // For caching
 
-	for _, key := range limits {
-		limit := key.(*egoscale.ResourceLimit)
+		err = forEachZone(allZones, func(zone string) error {
+			ctx := exoapi.WithEndpoint(gContext, exoapi.NewReqEndpoint(gCurrentAccount.Environment, zone))
 
-		if used, err := fetchUsedResources(limit.ResourceTypeName); used != -1 {
+			instances, err := cs.ListInstances(ctx, zone)
 			if err != nil {
-				return nil, err
+				return fmt.Errorf("unable to list Compute instances: %s", err)
+			}
+			curUsage.Store(limitGPUs, 0)
+			cur, _ := curUsage.LoadOrStore(limitComputeInstances, 0)
+			curUsage.Store(limitComputeInstances, cur.(int)+len(instances))
+
+			for _, instance := range instances {
+				instanceType, cached := instanceTypes[instance.InstanceTypeID]
+				if !cached {
+					instanceType, err = cs.GetInstanceType(ctx, zone, instance.InstanceTypeID)
+					if err != nil {
+						return fmt.Errorf(
+							"unable to retrieve Compute instance type %q: %s",
+							instance.InstanceTypeID,
+							err)
+					}
+					instanceTypes[instance.InstanceTypeID] = instanceType
+				}
+
+				if strings.HasSuffix(instanceType.Family, "gpu") {
+					cur, _ = curUsage.Load(limitGPUs)
+					curUsage.Store(limitGPUs, cur.(int)+int(instanceType.GPUs))
+				}
```

Comment:
AFAIU it means that a running GPU instance is counted twice in limits: for instance limits and GPU limits, right?

Reply:
Answering to myself by practicing: Yes

```
❯ gco ch30748
Branch 'ch30748' set up to track remote branch 'ch30748' from 'origin' by rebasing.
Switched to a new branch 'ch30748'
❯ make build
[...]
❯ ./bin/exo config
✔ prd*
❯ ./bin/exo limits
┼────────────────────────┼──────┼─────┼
│        RESOURCE        │ USED │ MAX │
┼────────────────────────┼──────┼─────┼
│ IP addresses           │ 0    │ 5   │
│ IAM API keys           │ 3    │ 100 │
│ GPUs                   │ 0    │ 1   │
│ Private Networks       │ 0    │ 32  │
│ Instances              │ 1    │ 20  │
│ Templates              │ 0    │ 30  │
│ Network Load Balancers │ 0    │ 5   │
│ Snapshots              │ 0    │ 30  │
│ SKS clusters           │ 0    │ 3   │
│ SOS buckets            │ 1    │ 100 │
┼────────────────────────┼──────┼─────┼
❯ ./bin/exo vm create gpu-001 -o GPU-small
Creating single-use SSH key
Deploying "gpu-001" ⠏                                                                                done
[...]
❯ ./bin/exo limits
┼────────────────────────┼──────┼─────┼
│        RESOURCE        │ USED │ MAX │
┼────────────────────────┼──────┼─────┼
│ IP addresses           │ 0    │ 5   │
│ IAM API keys           │ 3    │ 100 │
│ GPUs                   │ 1    │ 1   │  <-- GPU count incremented as expected
│ Private Networks       │ 0    │ 32  │
│ Instances              │ 2    │ 20  │ <-- Instances count incremented too
│ Templates              │ 0    │ 30  │
│ Network Load Balancers │ 0    │ 5   │
│ Snapshots              │ 0    │ 30  │
│ SKS clusters           │ 0    │ 3   │
│ SOS buckets            │ 1    │ 100 │
┼────────────────────────┼──────┼─────┼
```

Reply:
@zyegfryed indeed, `GPUs` refers here to the number of GPU cards utilized in Compute instances within the organization, it is not a *type* of Compute instance.